### Working with images
##### Check that we have correct libraries installed

In [0]:
# Working with Images
import tensorflow
import keras
import h5py

##### Import image into dataframe

In [0]:
# Import Images to DataFrame

img_path = '/FileStore/tables/typewriter.jpg'

In [0]:
df = spark.read.format('image').load(img_path)

In [0]:
df.printSchema()

root
-- image: struct (nullable = true)
 |-- origin: string (nullable = true)
 |-- height: integer (nullable = true)
 |-- width: integer (nullable = true)
 |-- nChannels: integer (nullable = true)
 |-- mode: integer (nullable = true)
 |-- data: binary (nullable = true)

##### Using pre-trained models

##### How this deep image predictor works?
- once the image is input, it needs to determine which of 1,000 classes of ImageNet the object and the image corresponds to.
- by default, each of the objects is given a probability.
- so that the deep image predictor sorts them based on probability.
- since we specified topK=5, it displays the object with top five highest probabilities.

In [0]:
from sparkdl import DeepImagePredictor

In [0]:
# input will be 'image' column of df 
# decodePredictions: making sense of the class from that sysnet file (the one we will be using)
# topK 5: we want to see top 5 predictions.

predictor = DeepImagePredictor(inputCol='image', outputCol='predicted_labels', modelName='ResNet50', decodePredictions=True, topK=5)

In [0]:
# transform the da`taframe
predictor_df = predictor.transform(df)

INFO:tensorflow:Froze 320 variables.
INFO:tensorflow:Converted 320 variables to const ops.
INFO:tensorflow:Froze 0 variables.
INFO:tensorflow:Converted 0 variables to const ops.

In [0]:
# get the prediction
predictor_df.select('predicted_labels').show(truncate=False)

+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
predicted_labels |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
[[n04505470, typewriter_keyboard, 0.97282547], [n04264628, space_bar, 0.02593652], [n03085013, computer_keyboard, 4.2348055E-4], [n03187595, dial_telephone, 1.5519575E-4], [n03075370, combination_lock, 8.707535E-5]]|
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

We can see that the top result is "Typewrite Keyboard" with probability with 0.97282547, which is accurate. This pretained model ResNet50 is working well.